modules/adServerModule/adServer.js

```diff
@@ -0,0 +1,87 @@
+import { ParentModule, SubmoduleBuilder } from '../shared/parentModule.js';
+import { module } from '../../src/hook.js';
+
+/**
+ * Ad Server Provider interface. All submodules of Ad Server Core must adhere to this.
+ * @typedef {Object} AdServerProvider
+ * @function getAdTagUrl - Builds an ad tag URL for a Bid.
+ */
+
+/**
+ * @function AdServerProvider#getAdTagUrl
+ * @param {Object} adUnit - the Ad Unit associated to the bid
+ * @param {string} baseAdTagUrl - Ad Tag onto which the targeting params of the bid can be appended.
+ * @returns {string}
+ */
+
+/**
+ * Config used to indicate which Ad Server to use to obtain the video ad tag url.
+ * @typedef {Object} AdServerConfig
+ * @property {string} vendorCode - The identifier of the AdServer vendor (i.e. gam, adloox, etc).
+ * @property {string} baseAdTagUrl - Ad Tag onto which the targeting params of the bid can be appended.
+ * @property {Object|undefined} params: Optional configuration block specific to the Ad Server vendor.
+ */
+
+/**
+ * Routes commands to the appropriate Ad Server Submodule
+ * @typedef {Object} AdServerCore
+ * @function registerAdServer
+ * @function getAdTagUrl
+ */
+
+const adServerDirectory = {};
+
+/**
+ * @constructor
+ * @param {ParentModule} parentModule_
+ * @returns {AdServerCore}
+ */
+export function AdServerCore(parentModule_) {
+  const parentModule = parentModule_;
+
+  /**
+   * @name AdServerCore#registerAdServer
+   * @param {AdServerConfig} config
+   */
+  function registerAdServer(config) {
+    const vendorCode = config.vendorCode;
+    try {
+      parentModule.registerSubmodule(vendorCode, vendorCode, config);
+    } catch (e) {}
+  }
+
+  /**
+   * Builds an ad tag URL for a Bid
+   * @name AdServerCore#getAdTagUrl
+   * @param {string} vendorCode - identifier of the Ad Server Provider type i.e. GAM
+   * @param {Object} adUnit - the Ad Unit associated to the bid
+   * @param {string} baseAdTagUrl - Ad Tag onto which the targeting params of a bid can be appended
+   */
+  function getAdTagUrl(vendorCode, adUnit, baseAdTagUrl) {
+    const submodule = parentModule.getSubmodule(vendorCode);
+    return submodule && submodule.getAdTagUrl(adUnit, baseAdTagUrl);
+  }
+
+  return {
+    registerAdServer,
+    getAdTagUrl
+  }
+}
+
+/**
+ * @function coreAdServerFactory
+ * @summary Factory to create an instance of Core Ad Server
+ * @returns {AdServerCore}
+ */
+export function coreAdServerFactory() {
+  const adServerSubmoduleBuilder = SubmoduleBuilder(adServerDirectory);
+  const parentModule = ParentModule(adServerSubmoduleBuilder);
+  const adServerCore = AdServerCore(parentModule);
+  return adServerCore;
+}
+
+function attachAdServerFactory(submoduleFactory) {
+  adServerDirectory[submoduleFactory.vendorCode] = submoduleFactory;
+}
+
+module('adServer', attachAdServerFactory)
```

Comment:
I still think we'd be better off delaying the adserver interface / directory and doing that separately; it would simplify this PR a lot.

At the moment, because there's no `index.js` in `adServerModule`, the dependency here is on `videoModule` (that's what `import`s the `module('adServer')` call). If you try to separate that out, you'll have the problem of trying to make this depend on `adServerModule`, which should depend on `shared`, which will be difficult to juggle without any `import`s. 

If you remove this, `shared` would only be used by `videoModule` and could be moved there which would allow you to keep most of the `import`s from it.

Reply:
I'll expand on this a bit - a complete ad server interface, when we get there, will almost certainly need a lot of support from core. There are many ugly instances where core has special cases for GAM that would need to be captured in it (see https://github.com/prebid/Prebid.js/issues/7638).

So most likely the future complete version of your `adServerModule` would need to be included in core, with specific ad server submodules filling in the blanks - it's unlikely that much the work you'd have to do here to keep it out of core (and removing all the `import`s) could be reused. 

But at the same time, I don't think the problem is well defined enough yet to justify asking you to make this more future-proof. Hence why I think it's better to sidestep this completely for now.